### PR TITLE
New docker images to use in our test-suite

### PR DESCRIPTION
**Pull request by jgsogo**
ci-unittests
-----------

This image contains only Python, **no system tools are available (there is no git, cmake or compiler)**.
It provides several python versions that can be referenced using environment variables like PY27,
PY35, PY37,...

ci-functional
-------------

This image contains the same Python versions as `ci-unittests`, but it also contains **system tools**
like Git, CMake, GCC or Clang. The installed versions are available using environment variables and
they can also be installed like the preferred alternative when starting the container:

```
docker run -it travis-ci-ci-functional:latest gcc 7 clang 9 cmake 3.16.4
```

### CODE REVIEW

File: docker_images/ci-functional/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,84 @@
+FROM ubuntu:eoan
+
+LABEL maintainer="Conan.io <info@conan.io>"
+
+ARG DEBIAN_FRONTEND=noninteractive
+
+ENV PY27 /root/.pyenv/versions/2.7.17/bin/python
```
Comment (uilianries) [inline docker_images/ci-functional/Dockerfile]: keep all env vars on same docker layer, it helps keeping the docker image smaller.

    ENV PY35 /root/.pyenv/versions/3.5.9/bin/python \
        PY37 /root/.pyenv/versions/3.7.6/bin/python \
        ...

File: docker_images/ci-functional/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+ENV GCC_9 /usr/bin/gcc-9
+ENV GXX_9 /usr/bin/g++-9
+
+ENV CLANG_7 /usr/bin/clang-7
+ENV CLANGXX_7 /usr/bin/clang++-7
+ENV CLANG_9 /usr/bin/clang-9
+ENV CLANGXX_9 /usr/bin/clang++-9
+
+
+RUN apt-get -qq update \
```
Comment (uilianries) [inline docker_images/ci-functional/Dockerfile]: Same here, after concluding your development stage, put all on same RUN statement.

File: docker_images/ci-functional/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    && pyenv install 3.7.6 \
+    && pyenv install 3.5.9 \
+    && pyenv install 2.7.17
+
+# Install CMake versions (using Conan) and register alternatives
+RUN $PY38 -m venv ./conan_tmp && cd ./conan_tmp \
+    && ./bin/pip install conan \
+    && export CONAN_USER_HOME=/root/conan_packages/ \
+    && ./bin/conan install cmake/3.16.4@ \
+    && ./bin/conan install cmake/3.16.3@ \
```
Comment (uilianries) [inline docker_images/ci-functional/Dockerfile]: 2 versions?

File: docker_images/ci-functional/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+# Alternatives for Clang
+COPY select_clang.sh /root/select_clang.sh
+RUN chmod +x /root/select_clang.sh &&\
+    update-alternatives --install /usr/bin/clang clang /usr/bin/clang-9 10 &&\
+    update-alternatives --install /usr/bin/clang clang /usr/bin/clang-7 20 &&\
+    update-alternatives --install /usr/bin/clang++ clang++ /usr/bin/clang++-9 10 &&\
+    update-alternatives --install /usr/bin/clang++ clang++ /usr/bin/clang++-7 20
+
+COPY select_alternatives.sh /root/select_alternatives.sh
+RUN chmod +x /root/select_alternatives.sh
```
Comment (uilianries) [inline docker_images/ci-functional/Dockerfile]:
```suggestion
RUN chmod +x /root/select_alternatives.sh
RUN rm -rf /var/lib/apt/lists/*
```

Exclude APT cache for smaller docker image

File: docker_images/ci-unittests/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,32 @@
+FROM ubuntu:eoan
+
+LABEL maintainer="Conan.io <info@conan.io>"
+
+ARG DEBIAN_FRONTEND=noninteractive
+
+ENV PY27 /root/.pyenv/versions/2.7.17/bin/python
```
Comment (uilianries) [inline docker_images/ci-unittests/Dockerfile]:
```suggestion
ENV PY27 /root/.pyenv/versions/2.7.17/bin/python \
    PY35 /root/.pyenv/versions/3.5.9/bin/python \
    PY37 /root/.pyenv/versions/3.7.6/bin/python \
    PY37 /root/.pyenv/versions/3.7.6/bin/python
```

File: docker_images/ci-unittests/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    && export PATH="/root/.pyenv/bin:$PATH" \
+    && pyenv init - \
+    && pyenv virtualenv-init - \
+    && pyenv update \
+    && pyenv install 3.8.1 \
+    && pyenv install 3.7.6 \
+    && pyenv install 3.5.9 \
+    && pyenv install 2.7.17
+
+RUN apt-get -qq remove make build-essential git gcc
```
Comment (uilianries) [inline docker_images/ci-unittests/Dockerfile]:
```suggestion
    && apt-get -qq remove make build-essential git gcc \
    && rm -rf /var/lib/apt/lists
```

File: docker_images/ci-unittests/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    \
+    && curl -L https://github.com/pyenv/pyenv-installer/raw/master/bin/pyenv-installer | bash \
+    && export PATH="/root/.pyenv/bin:$PATH" \
+    && pyenv init - \
+    && pyenv virtualenv-init - \
+    && pyenv update \
+    && pyenv install 3.8.1 \
+    && pyenv install 3.7.6 \
+    && pyenv install 3.5.9 \
+    && pyenv install 2.7.17
```
Comment (uilianries) [inline docker_images/ci-unittests/Dockerfile]:
```suggestion
    && pyenv install 2.7.17 \
```

File: docker_images/ci-functional/Dockerfile (hunk 3 above)
Comment (jgsogo) [reply docker_images/ci-functional/Dockerfile]: I'd like to add more than two versions, we may need to test Conan against different CMake versions, but right now we only have `3.16` is Conan Center 😞 

There are some meaningful versions we would like to test, Conan uses some features and sometimes we are not so good implementing that functionality with older versions.

Comment (jgsogo) [reply docker_images/ci-functional/Dockerfile]: Of course, using here `3.16.3` and `3.16.4` is just an example

**Comment by jgsogo** [comment]
Thanks @uilianries . I'm following your advice trying to reduce (to not make bigger) the size of the images (in this PR before merging it: https://github.com/jgsogo/conan_ci_jenkins/pull/3). 

I've read that 
> In Docker 1.10 and higher, only RUN, COPY, and ADD instructions create layers. Other instructions create temporary intermediate images, and no longer directly increase the size of the build.
so I'll keep the `ENV` calls.

But I prefer to think about this as a first step, there will be tons of opportunities to reduce their size.

File: docker_images/ci-functional/Dockerfile (hunk 3 above)
Comment (uilianries) [reply docker_images/ci-functional/Dockerfile]: But why, I mean, do you want to check behavior? just curious

**Comment by uilianries** [comment]
Anyway, this PR looks a huge improvement! :clap:

File: docker_images/ci-functional/Dockerfile (hunk 3 above)
Comment (jgsogo) [reply docker_images/ci-functional/Dockerfile]: Eventually I would like to run the Conan test suite for different CMake versions, to check we do not rely on any new/old behavior. Have a look at these "docs" here: https://github.com/conan-io/conan_ci_jenkins/pull/10

Comment (uilianries) [reply docker_images/ci-functional/Dockerfile]: nice!

Comment (jgsogo) [reply docker_images/ci-functional/Dockerfile]: Maybe now you understand my comment here: https://github.com/conan-io/conan-center-index/pull/633#issuecomment-573240765 😆

**Pull request merged**